src/workerd/api/gpu/gpu-device.c++

```diff
@@ -287,8 +262,166 @@ GPUDevice::createComputePipeline(GPUComputePipelineDescriptor descriptor) {
+          c->fulfiller->fulfill(kj::mv(err));
+          break;
+        }
+        case WGPUErrorType::WGPUErrorType_Validation: {
+          jsg::Ref<GPUError> err =
+              jsg::alloc<GPUValidationError>(kj::str(message));
+          c->fulfiller->fulfill(kj::mv(err));
+          break;
+        }
+        case WGPUErrorType::WGPUErrorType_Unknown:
+        case WGPUErrorType::WGPUErrorType_DeviceLost:
+          c->fulfiller->reject(JSG_KJ_EXCEPTION(FAILED, TypeError, message));
+          break;
+        default:
+          c->fulfiller->reject(
+              JSG_KJ_EXCEPTION(FAILED, TypeError, "unhandled error type"));
+          break;
+        }
+      },
+      ctx);
+
+  auto &context = IoContext::current();
+  return context.awaitIo(kj::mv(paf.promise));
+}
+
+jsg::Promise<jsg::Ref<GPUComputePipeline>>
+GPUDevice::createComputePipelineAsync(GPUComputePipelineDescriptor descriptor) {
+  wgpu::ComputePipelineDescriptor desc =
+      parseComputePipelineDescriptor(descriptor);
+
+  struct Context {
+    kj::Own<kj::PromiseFulfiller<jsg::Ref<GPUComputePipeline>>> fulfiller;
+    AsyncTask task;
+  };
+  auto paf = kj::newPromiseAndFulfiller<jsg::Ref<GPUComputePipeline>>();
+  // This context object will hold information for the callback, including the
+  // fullfiller to signal the caller with the result, and an async task that
+  // will ensure the device's Tick() function is called periodically. It will be
+  // deallocated at the end of the callback function.
+  auto ctx = new Context{kj::mv(paf.fulfiller), AsyncTask(kj::addRef(*async_))};
+
+  device_.CreateComputePipelineAsync(
+      &desc,
+      [](WGPUCreatePipelineAsyncStatus status, WGPUComputePipeline pipeline,
+         char const *message, void *userdata) {
+        // Note: this is invoked outside the JS isolate lock
+        auto c = std::unique_ptr<Context>(static_cast<Context *>(userdata));
+
+        switch (status) {
+        case WGPUCreatePipelineAsyncStatus::
+            WGPUCreatePipelineAsyncStatus_Success:
+          c->fulfiller->fulfill(
+              jsg::alloc<GPUComputePipeline>(kj::mv(pipeline)));
+          break;
+        default:
+          c->fulfiller->reject(
+              JSG_KJ_EXCEPTION(FAILED, TypeError, "unknown error"));
+          break;
+        }
+      },
+      ctx);
+
+  auto &context = IoContext::current();
+  return context.awaitIo(kj::mv(paf.promise));
+}
+
+jsg::Ref<GPUQueue> GPUDevice::getQueue() {
+  auto queue = device_.GetQueue();
+  return jsg::alloc<GPUQueue>(kj::mv(queue));
+}
+
+void GPUDevice::destroy() {
+  // TODO(soon): handle lost device promise when we have it
+  device_.Destroy();
+}
+
+GPUDevice::GPUDevice(wgpu::Device d)
+    : device_(d), async_(kj::refcounted<AsyncRunner>(d)) {
+  device_.SetLoggingCallback(
+      [](WGPULoggingType type, char const *message, void *userdata) {
+        KJ_LOG(INFO, "WebGPU logging", kj::str(type), message);
```

Comment:
What's the purpose of the `KJ_LOG` statements? We do not generate `INFO` logs in production so if these are meant to be seen in production they need to use ` KJ_LOG(WARNING, ...`

Reply:
This is purely for debugging purposes when someone is using `--verbose`. I don't yet have a place to send these.

Would it be possible/advisable to simulate `console.log()`s so that the script author could see these?